File: 31.下一个排列.py
#
# @lc app=leetcode.cn id=31 lang=python
#
# [31] 下一个排列
#

# @lc code=start
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        i = len(nums) - 2
        while i >= 0 and nums[i + 1] <= nums[i]:
            i -= 1
        if i >= 0:
            j = len(nums) - 1
            while j >= i and nums[j] <= nums[i]:
                j -= 1
            nums[i], nums[j] = nums[j], nums[i]
        nums[i+1:] = reversed(nums[i+1:])
        
# @lc code=end

File: 31.下一个排列.cpp
/*
 * @lc app=leetcode.cn id=31 lang=cpp
 *
 * [31] 下一个排列
 */

// @lc code=start
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if (nums.size() == 1) {
            return;
        }
        if (nums.size() == 2) {
            std::swap(nums[0], nums[1]);
            return;
        }

        int r = nums.size() - 1;
        while (r >= 1) {
            if (nums[r] <= nums[r - 1]) {
                r--;
            } else {
                break;
            }
        }
        if (r == 0) {
            std::reverse(nums.begin(), nums.end());
        } else {
            for (int i = nums.size() - 1; i >= r; --i) {
                if (nums[i] > nums[r-1]) {
                    std::swap(nums[i], nums[r-1]);
                    break;
                }
            }
           
            std::sort(nums.begin() + r, nums.end());
        }
    }
};
// @lc code=end

